selfdrive/manager.py

```diff
@@ -478,12 +500,40 @@ def manager_prepare(spinner=None):
   os.chdir(os.path.dirname(os.path.abspath(__file__)))
 
   # Spinner has to start from 70 here
-  total = 100.0 if prebuilt else 50.0
+  total = 100.0 if prebuilt else 70.0
+
+  for retry in [True, False]:
+    progress = total
+    prep_failed = False
+    for i, p in enumerate(managed_processes):
+      e = prepare_managed_process(p)
+      progress = (100.0 - total) + total * (i + 1) / len(managed_processes)
+      if spinner is not None:
+        if e is None:
+          spinner.update("%d" % progress)
+        else:
+          prep_failed = True
+          for _ in range(10):
+            spinner.update("%d" % progress, format_spinner_error(str(e)))
+            time.sleep(1)
+          break
 
-  for i, p in enumerate(managed_processes):
-    if spinner is not None:
-      spinner.update("%d" % ((100.0 - total) + total * (i + 1) / len(managed_processes),))
-    prepare_managed_process(p)
+    if prep_failed:
+      if retry:
+        print("preparation failed, hard resetting in")
+        for i in range(5):
+          print(5 - i)
+          spinner.update("%d" % progress, "preparation failed, hard resetting in {}...".format(5 - i))
+          time.sleep(1)
+
+        subprocess.check_output(["git", "fetch"], cwd="/data/openpilot", stderr=subprocess.STDOUT, encoding='utf8')
+        r = subprocess.check_output(["git", "reset", "--hard", "@{u}"], cwd="/data/openpilot", stderr=subprocess.STDOUT,
+                                    encoding='utf8')
```

Comment:
Should more be here to let the user know they should connect to the internet before it fetches, and wait for that? Or should the reset be good enough? `updated` isn't running here so if a fork's owner writes bad code it would fail to start and pull the fix by itself.

Reply:
Let's remove the git reset for now. Not sure yet how I feel about doing that automatically.

Reply:
So just show the error for 10 seconds and raise exception for now?

Reply:
yes